### PR TITLE
Feature: Enforce OAuth2

### PR DESCRIPTION
**Pull request by andy89923**
## Description

This pull request let NRF service producer to use OAuth2 token to authorize. 
Also, when NF Register to NRF, the NRF would let NF knows whether the OAuth2 is required via NFProfile. 

## Usage
To enable oauth, modify the NRF config file. 
 ```yaml
configuration:
  sbi:
    oauth: true
```

### CODE REVIEW

**Review by ianchen0119 — APPROVED** [review]
LGTM!

**Comment by andy89923** [comment]
Pending for https://github.com/free5gc/openapi/pull/18.  Update for ```go.mod``` and ```go.sum```are required.

**Comment by tim-ywliu** [comment]
@andy89923 rebase and update openapi hash

File: internal/context/context.go [hunk 1]
```diff
@@ -135,24 +135,33 @@ func makeDir(filePath string) error {
...
+		if err != nil {
+			return errors.Wrapf(err, "Generate Error")
+		}
+
 		// Get NF's Public key from file
 		var nfPubKey *rsa.PublicKey
-		nfPubKey, err = openapi.ParsePublicKeyFromPEM(nfCertPath)
+		nfPubKey, err = oauth.ParsePublicKeyFromPEM(nfCertPath)
 		if err != nil {
 			return errors.Wrapf(err, "sign NF cert")
```
Comment (tim-ywliu) [inline internal/context/context.go:159]: line 150 `oauth.GenerateRSAKeyPair(nfCertPath, "")` should be moved here (when ParsePublicKayFromPEM failed)

File: internal/sbi/producer/access_token.go [hunk 2]
```diff
@@ -152,8 +155,13 @@ func AccessTokenScopeCheck(req models.AccessTokenReq) *models.AccessTokenErr {
 	}
 	if _, err = nfCert.Verify(opts); err != nil {
 		logger.AccTokenLog.Errorln("Certificate verify error: " + err.Error())
-		return &models.AccessTokenErr{
-			Error: "invalid_client",
+		// DEBUG
+		// In testing environment, this would leads to follwing error:
+		// certificate verify error: x509: certificate signed by unknown authority
+		if !strings.Contains(err.Error(), "unknown authority") {
```
Comment (tim-ywliu) [inline internal/sbi/producer/access_token.go:161]:
```
if strings.Contains(err.Error(), "unknown authority") {
    logger.AccTokenLog.Warnf("Certificate verify: %v", err)
} else {
    return &models.AccessTokenErr{
        Error: "invalid_client",
    }
}
```

Comment (andy89923) [reply internal/sbi/producer/access_token.go:161]: > ```
> if strings.Contains(err.Error(), "unknown authority") {
>     logger.AccTokenLog.Warnf("Certificate verify: %v", err)
> } else {
>     return &models.AccessTokenErr{
>         Error: "invalid_client",
>     }
> }
> ```

Would you prefer this adjustment, or do you feel the need for any modifications?

File: internal/sbi/producer/access_token.go [hunk 3]
```diff
@@ -152,8 +155,13 @@ func AccessTokenScopeCheck(req models.AccessTokenReq) *models.AccessTokenErr {
 	}
 	if _, err = nfCert.Verify(opts); err != nil {
 		logger.AccTokenLog.Errorln("Certificate verify error: " + err.Error())
```
Comment (tim-ywliu) [inline internal/sbi/producer/access_token.go:157]: remove this line

File: internal/sbi/producer/access_token.go (hunk 2 above)
Comment (tim-ywliu) [reply internal/sbi/producer/access_token.go:161]:
```
if strings.Contains(err.Error(), "unknown authority") {
    logger.AccTokenLog.Warnf("Certificate verify: %v", err)
} else {
    logger.AccTokenLog.Errorf("Certificate verify: %v", err)
    return &models.AccessTokenErr{ Error: "invalid_client", }
}
```

File: internal/context/context.go [hunk 4]
```diff
@@ -135,24 +135,32 @@ func makeDir(filePath string) error {
 	return nil
 }
 
-func SignNFCert(nfType, nfId string) error {
-	nfCertPath := openapi.GetNFCertPath(factory.NrfConfig.GetCertBasePath(), nfType)
+func SignNFCert(nfType, nfId, nfCertPem string) error {
```
Comment (tim-ywliu) [inline internal/context/context.go:138]: what is `nfCertPem` used for?

Comment (andy89923) [reply internal/context/context.go:138]: It is not used now. Would be removed in next commit.

**Review by tim-ywliu — APPROVED** [review]

**Pull request merged**